File: services/api/app/schemas/paar_szenen.py
"""Pydantic-Schemas für Beziehungsszenen im Paarraum."""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from app.services.paar_szenen_katalog import MAX_KOMMENTAR

Art = Literal["getrennt", "geraten"]


# ── Das Regal ────────────────────────────────────────────────────────────────
class RegalEintrag(BaseModel):
    scene_slug: str
    title: str | None = None
    perspective: str | None = None
    wirkungen: list[str] = Field(default_factory=list)
    grund: str | None = None
    verwaist: bool = False


class RegalAntwort(BaseModel):
    """Beide Auswahlen nebeneinander.

    ``gemeinsam`` ist der Punkt der Übung: Zwei Menschen, die sich über nichts einig sind,
    sind sich fast immer über eine Szene einig — und das ist ein besserer Anfang als jede
    Frage nach dem Problem.
    """

    meine: list[RegalEintrag]
    ihre: list[RegalEintrag]
    gemeinsam: list[str]
    max: int
    empfohlen: int


class RegalWahl(BaseModel):
    scene_slug: str
    grund: str | None = Field(default=None, max_length=500)


# ── Die Runde ────────────────────────────────────────────────────────────────
class RundeVorschlagen(BaseModel):
    scene_slug: str
    art: Art
    #: Die Frage, die von der Geschichte auf das Paar zeigt. Abwählbar, weil ihre Antwort
    #: sich als Vorwurf lesen lässt.
    mit_bruecke: bool = True


class Antworten(BaseModel):
    """Roh, weil die Form von der Art der Runde abhängt.

    Bei ``getrennt`` steht je Frage ein Text, bei ``geraten`` ein Paar aus eigener Antwort
    und Vermutung. Geprüft wird serverseitig gegen den Fragenkatalog
    (:func:`paar_szenen_katalog.bereinigen`) — was dort nicht hineinpasst, fällt weg.
    """

    antworten: dict[str, Any] = Field(default_factory=dict)
    kommentar: str | None = Field(default=None, max_length=MAX_KOMMENTAR)


class TrefferZeile(BaseModel):
    frage: str
    vermutet: str
    wirklich: str
    getroffen: bool


class RundeAnsicht(BaseModel):
    """Was eine Person von der Runde sehen darf.

    ``ihre_antworten`` ist leer, solange nicht aufgedeckt ist — das entscheidet der Dienst,
    nicht die Oberfläche. ``sie_ist_fertig`` ist die einzige Auskunft über die andere Person
    davor, und sie ist nötig: Ohne sie wartet man vor einem Bildschirm, der nichts sagt.
    """

    id: str
    scene_slug: str
    title: str | None = None
    perspective: str | None = None
    art: Art
    mit_bruecke: bool
    status: Literal["vorgeschlagen", "laeuft", "aufgedeckt", "abgelehnt"]
    ich_habe_vorgeschlagen: bool
    #: Kommt vom Server, damit Wortlaut und Optionen nur an einer Stelle stehen.
    fragen: list[dict[str, Any]] = Field(default_factory=list)
    meine_antworten: dict[str, Any] = Field(default_factory=dict)
    ihre_antworten: dict[str, Any] = Field(default_factory=dict)
    ich_bin_fertig: bool = False
    sie_ist_fertig: bool = False
    treffer: list[TrefferZeile] = Field(default_factory=list)
    verwaist: bool = False


class PaarSzenenStand(BaseModel):
    regal: RegalAntwort
    runde: RundeAnsicht | None = None

File: apps/web/src/api/paarSzenen.ts
/**
 * Beziehungsszenen im Paarraum — Regal und Runde.
 *
 * **Was hier auffällt und Absicht ist:** Es gibt keine Methode, die die Antworten der
 * anderen Person abruft. Vor dem Aufdecken gibt der Server sie nicht heraus, und danach
 * stehen sie im Rundenstand. Ein eigener Endpunkt dafür wäre ein zweiter Weg an der
 * Blindheit vorbei.
 */
import { apiClient } from './client'

export type PaarSzenenArt = 'getrennt' | 'geraten'
export type RundenStatus = 'vorgeschlagen' | 'laeuft' | 'aufgedeckt' | 'abgelehnt'

export interface RegalEintrag {
  scene_slug: string
  title: string | null
  perspective: string | null
  wirkungen: string[]
  grund: string | null
  verwaist: boolean
}

export interface RegalAntwort {
  meine: RegalEintrag[]
  ihre: RegalEintrag[]
  /** Slugs, die beide gewählt haben — der Punkt der Übung. */
  gemeinsam: string[]
  max: number
  empfohlen: number
}

export interface PaarFrageOption { key: string; label: string }

/** Bei `getrennt` mit `hinweis`/`platzhalter`, bei `geraten` mit `optionen`/`vermutung`. */
export interface PaarFrage {
  key: string
  label: string
  hinweis?: string
  platzhalter?: string
  zeilen?: number
  vermutung?: string
  optionen?: PaarFrageOption[]
}

export interface TrefferZeile {
  frage: string
  vermutet: string
  wirklich: string
  getroffen: boolean
}

export interface RundeAnsicht {
  id: string
  scene_slug: string
  title: string | null
  perspective: string | null
  art: PaarSzenenArt
  mit_bruecke: boolean
  status: RundenStatus
  ich_habe_vorgeschlagen: boolean
  fragen: PaarFrage[]
  meine_antworten: Record<string, unknown>
  /** Leer, solange nicht aufgedeckt ist — das entscheidet der Server. */
  ihre_antworten: Record<string, unknown>
  ich_bin_fertig: boolean
  /** Die einzige Auskunft über sie vor dem Aufdecken. Ohne sie wartet man vor nichts. */
  sie_ist_fertig: boolean
  treffer: TrefferZeile[]
  verwaist: boolean
}

export interface PaarSzenenStand {
  regal: RegalAntwort
  runde: RundeAnsicht | null
}

const basis = (coupleId: string) => `/couple/links/${coupleId}/szenen`

export const paarSzenenApi = {
  stand: (coupleId: string) =>
    apiClient.get<PaarSzenenStand>(basis(coupleId)).then(r => r.data),

  regalWaehlen: (coupleId: string, scene_slug: string, grund?: string | null) =>
    apiClient
      .put<RegalAntwort>(`${basis(coupleId)}/regal`, { scene_slug, grund: grund ?? null })
      .then(r => r.data),

  regalEntfernen: (coupleId: string, slug: string) =>
    apiClient.delete<RegalAntwort>(`${basis(coupleId)}/regal/${slug}`).then(r => r.data),

  vorschlagen: (
    coupleId: string, scene_slug: string, art: PaarSzenenArt, mit_bruecke: boolean,
  ) =>
    apiClient
      .post<RundeAnsicht>(`${basis(coupleId)}/runden`, { scene_slug, art, mit_bruecke })
      .then(r => r.data),

  annehmen: (coupleId: string, roundId: string) =>
    apiClient
      .post<RundeAnsicht>(`${basis(coupleId)}/runden/${roundId}/annehmen`)
      .then(r => r.data),

  ablehnen: (coupleId: string, roundId: string) =>
    apiClient.post(`${basis(coupleId)}/runden/${roundId}/ablehnen`).then(() => undefined),

  antwortenSichern: (
    coupleId: string, roundId: string,
    antworten: Record<string, unknown>, kommentar?: string | null,
  ) =>
    apiClient
      .put<RundeAnsicht>(`${basis(coupleId)}/runden/${roundId}/antworten`, {
        antworten, kommentar: kommentar ?? null,
      })
      .then(r => r.data),

  fertig: (coupleId: string, roundId: string) =>
    apiClient
      .post<RundeAnsicht>(`${basis(coupleId)}/runden/${roundId}/fertig`)
      .then(r => r.data),
}
